### PR TITLE
JSON header property support for Auth Proxy

### PR DESCRIPTION
**Pull request by tairila**
Some API gateway components (e.g. following: https://github.com/nokia/kong-oidc ) add logged in user as a JSON string in a header instead of a separate header containing only the username. This addition allows the Auth Proxy to fetch username from the specified JSON key in a header.

### CODE REVIEW

**Comment by marefr** [comment]
How is this different to existing PR #13840?

**Comment by tairila** [comment]
Hi @marefr, Antti (my old colleague) is nowadays working for another employer and I think he is not interested to continue his pull request further, so that's why I have opened a new one as we need this change.

**Comment by DanCech** [comment]
This seems like it would be much more powerful if it followed the `headers` pattern of being able to map multiple items in the json to different properties ( I see @torkelo made a similar comment on #13840).

I envision the config looking like:
```
# If HTTP Header is a JSON string, set this to true. Defaults to 'false'.
json_header = true
# Needed when using json_header, specify the JSON properties that contain Login, Email and/or Name
json_properties = Login:preferred_username Email:email
```

**Review by bergquist — CHANGES_REQUESTED** [review]
Thank you for taking the time to contribute to Grafana.

Considering the current complexity of `initContextWithAuthProxy` we won't introduce any changes without tests that verify the change.

**Comment by tairila** [comment]
Hi @bergquist please review as test cases have been added over a month ago.

File: docs/sources/auth/auth-proxy.md [hunk 1]
```diff
@@ -25,6 +25,13 @@ enabled = true
 header_name = X-WEBAUTH-USER
 # HTTP Header property, defaults to `username` but can also be `email`
 header_property = username
+# Optionally define more headers to sync other user attributes
+# Example `headers = Name:X-WEBAUTH-NAME Email:X-WEBAUTH-EMAIL`
+headers =
+# If HTTP Header is a JSON string, set this to true. Defaults to 'false'.
+json_header = true
+# Needed when json_header is used, specifies the JSON key(s) for the user like username, email and/or full name.
+json_header_property = Login:preferred_username, Email:mail, Name:full_name
```
Comment (DanCech) [inline docs/sources/auth/auth-proxy.md]:
```suggestion
json_properties = Login:preferred_username, Email:mail, Name:full_name
```

Comment (tairila) [reply docs/sources/auth/auth-proxy.md]: Hi @DanCech, I would like to keep _header_ as a part of the name as then it is more clear that it goes together with json_header, so what would you say if I rename it to json_header_properties?

Comment (DanCech) [reply docs/sources/auth/auth-proxy.md]: I'm fine with that (and agree `json_header_properties` makes more sense), it just needs to be consistent everywhere, right now this is the only place that doesn't call it `json_properties`

Comment (tairila) [reply docs/sources/auth/auth-proxy.md]: That's true, so I will use "json_header_properties" everywhere.

**Comment by marefr** [comment]
@tairila we have a similar PR (#14683) open but for generic oauth which suggests using [JMESPath](http://jmespath.org/). I personally find this to be suitable for extracting auth proxy json properties as well, do you agree? 

However, not sure if the author of referenced PR is still up for finishing that so let me come back regarding that.

**Comment by tairila** [comment]
Hi @marefr, yes I agree.

**Comment by marefr** [comment]
@tairila we've now merged #14683 which added support for using [JMESPath](http://jmespath.org/) to lookup/select email address in OAuth authentication. Are you still up for working on this? 

If you are, I think similar solution as suggested in https://github.com/grafana/grafana/pull/15762#issuecomment-469345540, but named `json_properties_path` or `json_path` with support for JMESPath to select `name`, `email` and `login`. I think that should be supported at least. Let us know if you need any help.

**Comment by tairila** [comment]
Hi @marefr, our team still needs these changes (+ some other small one), but as you have refactored the code and there are those JMESPath things I suggest that this pull request will be closed and somebody from our team opens a new one later.

**Comment by marefr** [comment]
@tairila sure sound good. Thanks. Closing this

**Pull request closed without merging**